File: backend/app/utils/text_filter.py
"""
Text filtering utilities for production-ready responses.
"""

import re
import logging

logger = logging.getLogger(__name__)


def remove_thinking_content(text: str) -> str:
    """
    Remove <thinking> content from text for production use.

    This function filters out internal reasoning content that should not be
    exposed to end users in production environments.

    Args:
        text: The input text that may contain <thinking> tags

    Returns:
        The cleaned text with <thinking> content removed
    """
    if not text:
        return text

    # Remove <thinking>...</thinking> blocks (case insensitive, multiline)
    # This regex handles:
    # - Case insensitive tags
    # - Multiline content
    # - Nested content within thinking blocks
    # - Potential malformed tags
    cleaned_text = re.sub(
        r'<thinking>.*?</thinking>',
        '',
        text,
        flags=re.IGNORECASE | re.DOTALL
    )

    # Also remove any orphaned opening or closing thinking tags
    cleaned_text = re.sub(
        r'</?thinking>', '', cleaned_text, flags=re.IGNORECASE
    )

    # Clean up any excessive whitespace that may result from removing content
    cleaned_text = re.sub(r'\n\s*\n\s*\n', '\n\n', cleaned_text)
    cleaned_text = cleaned_text.strip()

    # Log if thinking content was found and removed (for debugging)
    if text != cleaned_text:
        original_len = len(text)
        cleaned_len = len(cleaned_text)
        logger.debug(
            f"Removed thinking content from text "
            f"(original length: {original_len}, cleaned length: {cleaned_len})"
        )

    return cleaned_text


def should_filter_thinking_content() -> bool:
    """
    Determine if thinking content should be filtered out.

    In production environments, thinking content should always be filtered.
    This function provides a centralized way to control this behavior.

    Returns:
        True if thinking content should be filtered, False otherwise
    """
    # Always filter thinking content for production use
    # This could be made configurable via environment variables if needed
    return True

File: frontend/src/utils/textFilter.ts
/**
 * Utility functions for filtering internal AI reasoning content from user-facing text
 */

/**
 * Removes <thinking> tags and their content from text
 * @param text - Input text that may contain thinking tags
 * @returns Text with thinking content removed
 */
export function removeThinkingContent(text: string): string {
  if (!text) return text;
  
  // Remove <thinking>...</thinking> blocks (including multiline content)
  return text.replace(/<thinking>[\s\S]*?<\/thinking>/gi, '').trim();
}

/**
 * Checks if text contains thinking content that should be filtered
 * @param text - Input text to check
 * @returns True if text contains thinking tags
 */
export function hasThinkingContent(text: string): boolean {
  if (!text) return false;
  return /<thinking>[\s\S]*?<\/thinking>/i.test(text);
}

/**
 * Determines if thinking content should be filtered based on environment
 * For now, always filter in frontend to ensure production safety
 * @returns True if thinking content should be filtered
 */
export function shouldFilterThinkingContent(): boolean {
  // Always filter thinking content in frontend for production safety
  return true;
}
